crates/turbo-tasks-fs/src/lib.rs

```diff
@@ -343,60 +374,99 @@ impl DiskFileSystem {
+                                                batched_invalidate_path_dir.insert(PathBuf::from(parent));
+                                            }
+                                        });
+                                    }
+                                    // A single event emitted with both the `From` and `To` paths.
+                                    EventKind::Modify(ModifyKind::Name(RenameMode::Both)) => {
+                                        // For the rename::both, notify provides an array of paths in given order
+                                        if let [source, destination, ..] = &paths[..] {
+                                            batched_invalidate_path_and_children.insert(source.clone());
+                                            if let Some(parent) = source.parent() {
+                                                batched_invalidate_path_dir.insert(PathBuf::from(parent));
+                                            }
+                                            batched_invalidate_path_and_children.insert(destination.clone());
+                                            if let Some(parent) = destination.parent() {
+                                                batched_invalidate_path_dir.insert(PathBuf::from(parent));
+                                            }
+                                            #[cfg(not(any(target_os = "macos", target_os = "windows")))]
+                                            batched_new_paths.insert(destination.clone());
+                                        } else {
+                                            println!("Rename event does not contain source and destination paths {:#?}", paths);
+                                        }
+                                    }
+                                    // [NOTE] Observing `ModifyKind::Metadata(MetadataKind::Any)` is not a mistake.
+                                    // PACK-2437 addresses some cases notify doesn't emit event to trigger hmr,
+                                    // cases like starting from a file have contents -> clear & save.
+                                    // v4 doesn't emit any debounced event for those,
+                                    // and v6 emits an event with `ModifyKind::Metadata(MetadataKind::Any)`
+                                    // it is unclear why these case are not considered as normal modify, anyway we use certain
+                                    // event to workaround.
+                                    EventKind::Any | EventKind::Modify(ModifyKind::Any | ModifyKind::Metadata(MetadataKind::Any)) => {
+                                        batched_invalidate_path.extend(paths.clone());
+                                        batched_invalidate_path_and_children.extend(paths.clone());
+                                        batched_invalidate_path_and_children_dir.extend(paths.clone());
```

Comment:
That seem to be a very extensive invalidation. I could imagine it over invalidating when e.g. metadata of a folder changes

Reply:
true, the unfortunate thing is it comes as `metadata::any`, so really no way to figure out if it's other arbitaray metadata change or content change in disguise. 

I created https://github.com/notify-rs/notify/issues/564 to get some clarification, based on those answer we may update codes later.

Reply:
Still waiting for confirmation from upstream, I came to think this is the limitation of the mac os's underlying behavior. Modified event comes intermittently for the same repro case, and it went down to the kernel's event flag itself doesn't contain such thing.

I wanted to compare how other implementation does, and then found libuv's impl:

It originally relied on `kFSEventStreamEventFlagItemModified` as same as what we expected -  https://github.com/libuv/libuv/blob/ea8db64559680b589196b1ae01152c35d2232c41/src/unix/fsevents.c#L150

and then it fixed to includes all of possible-event data including metadata - 
https://github.com/libuv/libuv/commit/73cf3600d75a5884b890a1a94048b8f3f9c66876#diff-e12fdb1f404f1c97bbdcc0956ac90d7db0d811d9fa9ca83a3deef90c937a486cR95-R99 
to emit `UV_CHANGE` event.

So this change might create additional excessive notify event for us, in my opinion it's not worsely efficient than other watch implementation.

Reply:
Could we skip the `_and_children` invalidation? That event on a folder doesn't need to invalidate all contained files and folders, does it?

Reply:
I think that might work, yes.